testdata/dnn/download_models.py

```diff
@@ -50,63 +60,115 @@ def verify(self):
         except Exception as e:
             print('  verify {}'.format(e))
 
-    def get(self):
+    def cache_copy(self, cache):
+        if not cache:
+            return False
+        try:
+            candidate = cache / self.filename
+            if candidate.is_file():
+                print('  cache {} -> {}'.format(candidate, self.filename))
+                self.prepare_file()
+                shutil.copy(candidate, self.filename)
+                return True
+        except Exception as e:
+            print('  cache {}'.format(e))
+
+    def get_sub(self, cache, arch):
+        print('** {}'.format(self.filename))
+
         if self.verify():
-            print('  hash match - skipping')
             return True
+        if self.cache_copy(cache):
+            return self.verify()
+        if arch.is_file():
+            self.extract(arch)
+            return self.verify()
 
-        basedir = os.path.dirname(self.filename)
-        if basedir and not os.path.exists(basedir):
-            print('  creating directory: ' + basedir)
-            os.makedirs(basedir, exist_ok=True)
+    def get(self, cache, arch=None):
+        print("* {}".format(m.name))
 
-        if self.archive or self.member:
-            assert(self.archive and self.member)
-            print('  hash check failed - extracting')
-            print('  get {}'.format(self.member))
-            self.extract()
+        # Sub elements - first attempt (cache)
+        if len(self.sub) > 0:
+            res = [m.get_sub(cache, self.filename) for m in self.sub]
+            if all(res):
+                return True
+
+        # File - exists or get from cache or download from internet
+        verified = False
+        if self.verify():
+            verified = True
+        elif self.cache_copy(cache):
+            pass
         elif self.url:
             print('  hash check failed - downloading')
             print('  get {}'.format(self.url))
             self.download()
-        else:
-            assert self.downloader
+        elif self.gdrive:
             print('  hash check failed - downloading')
-            sz = self.downloader(self.filename)
-            print('  size = %.2f Mb' % (sz / (1024.0 * 1024)))
+            print('  get {}'.format(self.gdrive))
+            self.download_gdrive()
+        else:
+            raise Exception(info="Invalid model record", name=self.name)
 
-        print(' done')
-        print(' file {}'.format(self.filename))
-        candidate_verify = self.verify()
-        if not candidate_verify:
+        if verified or self.verify():
+            # Sub elements - second attempt (extract)
+            if len(self.sub) > 0:
+                res = [m.get_sub(cache, self.filename) for m in self.sub]
```

Comment:
Am I right that if there is an archive in the cache (broken or valid) but with a different SHA from `expected` archive SHA, we expect script to fail?
```
da39a3ee5e6b4b0d3255bfef95601890afd80709  cache/ssd_mobilenet_v1_coco_11_06_217.tar.gz
```
```python
    Model(
        name='MobileNet-SSD (TensorFlow)',
        url='http://download.tensorflow.org/models/object_detection/ssd_mobilenet_v1_coco_11_06_2017.tar.gz',
        sha='a88a18cca9fe4f9e496d73b8548bfd157ad286e2',
        filename='ssd_mobilenet_v1_coco_11_06_217.tar.gz',
        sub=[
            Model(
                member='ssd_mobilenet_v1_coco_11_06_2017/frozen_inference_graph.pb',
                filename='ssd_mobilenet_v1_coco.pb',
                sha='aaf36f068fab10359eadea0bc68388d96cf68139'
            )
        ]),
```
```
$ python download.py -d dst -c cache

* MobileNet-SSD (TensorFlow)
** ssd_mobilenet_v1_coco.pb
  expect aaf36f068fab10359eadea0bc68388d96cf68139
  verify - file does not exist
  expect a88a18cca9fe4f9e496d73b8548bfd157ad286e2
  verify - file does not exist
  cache /home/dkurt/tmp/cache/ssd_mobilenet_v1_coco_11_06_217.tar.gz -> ssd_mobilenet_v1_coco_11_06_217.tar.gz
------------------- False
  expect a88a18cca9fe4f9e496d73b8548bfd157ad286e2
  actual da39a3ee5e6b4b0d3255bfef95601890afd80709
  renaming invalid file to ssd_mobilenet_v1_coco_11_06_217.tar.gz.invalid
* FAIL
FAILED: MobileNet-SSD (TensorFlow) - ssd_mobilenet_v1_coco_11_06_217.tar.gz
```

Reply:
In other words, expect script to fail but not try to download an archive to `dst` if cache copy failed?

Reply:
That is also usecase which I didn't test (broken cache). I tried to fix it by moving `verify` to the cache copying procedure and cleanup in case of error.
Thank you!